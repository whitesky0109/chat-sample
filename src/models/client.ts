import { UserPayload } from 'public/store/actions/userAction';
import { RoomPayload } from 'public/store/actions/roomAction';

export interface IMessage {
  roomId?: string;
  type: string;
  message: File | string;
}

export interface StoreState {
  user: UserPayload;
  room: RoomPayload;
}
