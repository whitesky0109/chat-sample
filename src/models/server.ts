export interface IMigration {
  migration_id: string;
  action: string;
  args: any[];
}

export interface IRepository {
  migrations(): IMigration[];
}

export interface IService {
  init(): Promise<any>;
}

export interface IMigrationLog {
  id: number;
  migration_id: string;
  action: string;
  success: boolean;
  error: string;
  timestamp: Date;
}

export interface IMigrationLogs {
  // tslint:disable-next-line: variable-name
  [migration_id: string]: IMigrationLog;
}

export interface IInvite {
  roomId: string;
  users: string[];
}

export interface IMessage {
  userId: string;
  type: string;
  message: ArrayBuffer | string;
  date: Date;
  isMy: boolean;
}

export interface IUser {
  [userId: string]: string | null;
}

export interface IRoomInfo {
  name: string;
  users: IUser;
  creator: string;
}

export interface IRoom {
  [roomId: string]: IRoomInfo;
}
