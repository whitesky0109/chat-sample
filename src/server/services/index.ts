import 'reflect-metadata';
// System Services
import {
    LoggerService,
    SocketManagerService,
} from './system';

// Chat Service
import {
    UserService,
    RoomService,
} from './chat';

export {
    SocketManagerService,
    LoggerService,
    UserService,
    RoomService,
};
