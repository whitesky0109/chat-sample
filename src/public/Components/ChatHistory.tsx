import moment from 'moment';
import React, { Component, ReactNode } from 'react';
const classnames = require('classnames');

// models
import { IMessage } from 'models/server';

// Components
import Message from './Message';

export interface Props {
  messages: IMessage[];
}

export default class ChatHistory extends Component<Props> {
  messagesEnd?: HTMLDivElement;

  renderMyData(username: string, date: string): ReactNode {
    return <div className="msg-data align-right">
      <span className="msg-time" >{date}</span>
      <span className="msg-name" >{username}</span>
    </div>;
  }

  renderOterData(username: string, date: string): ReactNode {
    return <div className="msg-data align-left">
      <span className="msg-name" >{username}</span>
      <span className="msg-time" >{date}</span>
    </div>;
  }

  renderMessage(key: number, message: IMessage): ReactNode {
    const { date, userId, isMy } = message;
    const dateStr = moment(date).format('YYYY년 MM월 DD일 (ddd) hh:mm:ss');
    const dataDom = message.isMy
      ? this.renderMyData(userId, dateStr)
      : this.renderOterData(userId, dateStr);
    const msgCls = classnames({
      'float-right': isMy,
      'float-left': !isMy,
    });

    return <li key={key} className="clearfix">
      {dataDom}

      <div className={`msg other-msg ${msgCls}`}>
        <Message {...message} />
      </div>
    </li >;
  }
  componentDidUpdate() {
    this.scrollToBottom();
  }

  scrollToBottom = () => {
    if (this.messagesEnd) {
      this.messagesEnd.scrollIntoView({ behavior: 'smooth' });
    }
  }

  render() {
    const { messages } = this.props;
    const messageRender = messages.map((message, index) => {
      return this.renderMessage(index, message);
    });

    const setMessageDiv = (el: HTMLDivElement) => { this.messagesEnd = el; };

    return <div className="chat-history">
      <ul>
        {messageRender}
      </ul>
      <div ref={setMessageDiv} />
    </div>;
  }
}
