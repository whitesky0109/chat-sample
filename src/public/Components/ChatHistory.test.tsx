import React from 'react';
import { shallow } from 'enzyme';
import ChatHistory from './ChatHistory';
import { IMessage } from 'models/server';
const moment = require('moment');

describe('ChatHistory', () => {
  let component: any = null;
  const messages: IMessage[] = [{
    date: new Date(0),
    isMy: true,
    message: 'my message',
    type: 'text',
    userId: 'userid',
  }];

  it('renders correctly', () => {

    component = shallow(<ChatHistory messages={messages} />);
  });
  it('사용자 이름 확인', () => {
    expect(component.find('.msg-name').text()).toBe(messages[0].userId);
  });

  it('우측 포지션 확인', () => {
    const { className } = component.find('.msg-data').props();
    const pos = className.split(' ')[1];
    expect(pos).toBe('align-right');
  });

  it('좌측 포지션 확인', () => {
    messages[0].isMy = false;
    component = shallow(<ChatHistory messages={messages} />);
    const { className } = component.find('.msg-data').props();
    const pos = className.split(' ')[1];
    expect(pos).toBe('align-left');
  });

  it('시간 확인', () => {
    const text = component.find('.msg-time').text();
    const dateStr = moment(messages[0].date).format('YYYY년 MM월 DD일 (ddd) hh:mm:ss');
    expect(text).toBe(dateStr);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

});
