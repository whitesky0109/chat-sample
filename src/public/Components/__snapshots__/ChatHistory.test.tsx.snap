// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatHistory matches snapshot 1`] = `
<div
  className="chat-history"
>
  <ul>
    <li
      className="clearfix"
      key="0"
    >
      <div
        className="msg-data align-left"
      >
        <span
          className="msg-name"
        >
          userid
        </span>
        <span
          className="msg-time"
        >
          1970년 01월 01일 (Thu) 09:00:00
        </span>
      </div>
      <div
        className="msg other-msg float-left"
      >
        <Message
          date={1970-01-01T00:00:00.000Z}
          isMy={false}
          message="my message"
          type="text"
          userId="userid"
        />
      </div>
    </li>
  </ul>
  <div />
</div>
`;
